New function for TPF pixel analysis: inspect_pixels()

astrobel opened the pull request:
Hi all,

In this PR, I'd like to introduce the `inspect_pixels()` function for Lightkurve TPFs. Modelled after [this old script](https://github.com/astrobel/chancealignments2/blob/master/4_pixels.py), `inspect_pixels()` produces a plot displaying either the light curves in each pixel, or their associated periodograms. The plots are displayed such that the pixels align with what you'd see using `interact()`.

Here's an example plot, showing the normalised light curves for KIC 2437317, as a result of this line of code:

```tpf.inspect_pixels(normalized=True, periodogram=False, aperture_mask='threshold', threshold=2)```

![image](https://user-images.githubusercontent.com/15132789/86306133-cc5bd080-bc56-11ea-9545-464b59bb712b.png)

Note also that the function plots an aperture mask, whether pipeline, threshold, or user-generated, and can take a threshold value as a kwarg.

There are still some things that need to be improved/tested, and I'd appreciate anyone's feedback & input:
- The pixel size to figure size ratio is fine for regular TPFs, but what about bright/saturated stars which have large TPFs?
- At the moment, the periodogram version takes a very long time to run. Any ideas for how to trim that down, or if it's possible?
- The normalized light curves often retain some long-term trends, making the periodograms slightly less useful at the moment.
- In my original version of this function, users have the option to limit the scale of the periodogram, so for example you might only want to see the first 10% of the frequency range up to the Nyquist frequency. Would people be interested in seeing that implemented here?

Thanks!

barentsen commented on lightkurve/targetpixelfile.py:
```diff
@@ -1130,6 +1131,124 @@ def cutout(self, center=None, size=5):
+                        else:
+                            pixel_list.append(lc)
+
+        if aperture_mask == 'pipeline':
+            mask = self.pipeline_mask
+        elif aperture_mask == 'threshold':
+            mask = self.create_threshold_mask(**kwargs)
+        elif aperture_mask == 'all' or aperture_mask == None:
+            mask = np.zeros((self.shape[1], self.shape[2]), dtype='bool')
+        else:
+            mask = aperture_mask
```
Good news: the class has a hidden method called `_parse_aperture_mask` to take care of this for you, so I think you can simplify as follows:
```suggestion
mask = self._parse_aperture_mask(aperture_mask)
```

barentsen commented on lightkurve/targetpixelfile.py:
```diff
@@ -1130,6 +1131,124 @@ def cutout(self, center=None, size=5):
+
+        with plt.style.context(style):
+            if title is None:
+                title = 'Target ID: {}, Quarter: {}'.format(self.targetid, self.quarter)
```
K2 and TESS data don't have a quarter number, so I suggest showing only the target id:
```suggestion
                title = f'Target ID: {self.targetid}'
```
(Pro-tip: now that Lightkurve is Python 3 only, we can use f-strings as shown above, which are awesome!)

barentsen commented on lightkurve/targetpixelfile.py:
```diff
@@ -1130,6 +1131,124 @@ def cutout(self, center=None, size=5):
         return self.__class__(newfits)
 
 
+    def inspect_pixels(self, ax=None, normalized=False, periodogram=False, aperture_mask=None, style='lightkurve', title=None, **kwargs):
```
Nitpicking: let's name the argument `normalize` instead of `normalized`, for consistency with the `LightCurve.plot()` methods which also accept a `normalize` argument.
```suggestion
    def inspect_pixels(self, ax=None, normalize=False, periodogram=False, aperture_mask=None, style='lightkurve', title=None, **kwargs):
```

barentsen commented on lightkurve/targetpixelfile.py:
```diff
@@ -1130,6 +1131,124 @@ def cutout(self, center=None, size=5):
         return self.__class__(newfits)
 
 
+    def inspect_pixels(self, ax=None, normalized=False, periodogram=False, aperture_mask=None, style='lightkurve', title=None, **kwargs):
```
Regarding the name: I like the idea of having every method that creates a plot start with the prefix `plot_`, so they are easy to discover.  For example, we have `plot_river()` in LightCurve and `plot_diagnostics()` in RegressionCorrector.  Following this philosophy, I wonder if e.g.  `plot_pixels()`, `plot_pixelseries()`, or `plot_pixelgrid()` would be a good name for this feature? What do you think?

astrobel replied:
Good point! I'm in favour of keeping it simple, so how about `plot_pixels()`?

barentsen commented:
@astrobel  To add a basic unit test for this feature, you can start by adding the following to the bottom of `lightkurve/tests/test_targetpixelfile.py`:
```python
def test_plot_pixels():
    tpf = KeplerTargetPixelFile(filename_tpf_one_center)
    tpf.inspect_pixels()
    plt.close('all')
```

Then, simply type `pytest -k test_plot_pixels` to verify that it all works.  (More guidance on running the tests is available here: https://docs.lightkurve.org/about/testing.html )

(Next, you can imagine extending the test to use optional arguments etc.)

barentsen commented:
Just for inspiration: I noticed that @noraeisner uses the following plotting style as part of a similar plot produced by the [LATTE](https://github.com/noraeisner/LATTE) package:

![image](https://raw.githubusercontent.com/noraeisner/LATTE/master/example/example_output/94986319_individual_pixel_LCs_0.png)

astrobel commented:
Thanks for all the feedback, @barentsen! I'm starting to work on tests now, so I'll keep you updated. I'm definitely going to pass kwargs to the periodogram.

As for running systematics removal on the pixels, it would certainly be useful for my purposes, and I'd be keen to hear if more people would like that as a feature! In that case, would we then remove normalization and replace it with this?

I love the pixel plot from LATTE. I'm going to look into implementing an option that allows users to toggle the colour scale background, as well as see if I can get the mask working as an outline instead of shading to accommodate that (whether or not the colour scale background is turned on).

ojhall94 commented:
Hey @astrobel!

This PR is awesome. I've had a little play around with it and am having a strange bug--- the plot doesn't appear in the same style as it does in your (very pretty) example at the top of this PR, and the `normalized` kwarg doesnt' seem to affect the appearance of the plot. Not sure if this is a true bug or a result of this PR being a WIP, but looking at the code I would expect the plot to follow the MPLSTYLE.

I don't think any changes have happened to lightkurve `master` branch since you branched off for this PR--- if you rebase this branch onto master (or merge master into this branch) I think it should fix the slow periodogram issue, which will make everything easier to test.

![image](https://user-images.githubusercontent.com/22291663/86573688-a68f3e00-bf6c-11ea-997f-2e485d0b3542.png)

astrobel commented:
Thanks @ojhall94! The plot's look is a good catch; that's both a bug and a WIP thing... I've been updating the plotting function to do some different stuff and it seems like when there's no aperture specified it treats every pixel as part of the aperture. Thankfully that's easily fixed!

(Though I can't work out why the y axis label seems to have disappeared... ?)

And `normalized` isn't working because a few hours ago I changed it to `normalize` as per [this commit](https://github.com/KeplerGO/lightkurve/pull/771/commits/86a31f3c24be0f78b79475997b5ee0ffd82df8b3).

astrobel commented:
Just a quick update on where I'm at with this PR! Following @barentsen's suggestion above (and inspired by @noraeisner's work, though I didn't look at the LATTE code, just the image), I've added an option to plot the flux (in frame zero - would people maybe prefer an average?) as a face colour to each pixel, in which case the flux/power will be plotted in white instead of black for visibility. I've also added a new-look aperture outline, which is the same whether or not the flux is plotted; plotting fluxes (but not power) in dots for legibility; and tidied up the figure size to make the pixels square. This is definitely going to make larger TPFs unwieldy, though, so I'd like to look more into scaling it!

![image](https://user-images.githubusercontent.com/15132789/86878587-93c36780-c12c-11ea-8b77-3911559f1df2.png)

The above plot is generated by running `tpf.plot_pixels(normalize=True, show_flux=True, aperture_mask='threshold')`. (Also note the new function name!)

I've also added the ability to pass kwargs to `to_periodogram`. Previously the kwargs were for specifying a threshold, but `_parse_aperture_mask` is unaware of changing the threshold by kwarg, so that keyword isn't needed anymore. Now, for example, you can use the `nyquist_factor` keyword to only plot up to a certain fraction of the Nyqust frequency:

![image](https://user-images.githubusercontent.com/15132789/86883818-fb31e500-c135-11ea-9c4e-0655d699da9e.png)

... from `tpf.plot_pixels(periodogram=True, show_flux=True, aperture_mask='threshold', nyquist_factor=0.5)`.

I've also begun to implement some tests, which you can see in the latest commit.

danhey commented:
Hey Isabel, awesome work! I'm a huge fan of this PR. If I can put in my 2 cents, I think a cool feature would be to allow passing a corrector function which would apply to each light curve generated from the pixel. A good example of this is [here](https://github.com/KeplerGO/lightkurve/blob/master/lightkurve/collections.py#L101).

astrobel commented:
@danhey Corrector functions are here! Let me know if you (or anyone!) get a chance to play around with it, as I've only just tested some basic usage now.

barentsen commented on lightkurve/tests/test_targetpixelfile.py:
```diff
@@ -624,3 +624,16 @@ def test_get_header():
         tpf.header
+
+def test_plot_pixels():
+    tpfs = [KeplerTargetPixelFile(filename_tpf_one_center), TessTargetPixelFile(filename_tpf_one_center)]
```
The statement `TessTargetPixelFile(filename_tpf_one_center)` will yield this warning during the tests:
```LightkurveWarning: A Kepler data product is being opened using the `TessTargetPixelFile` class.```

While the warning could be silenced, it would be fine to simply run this test just for `KeplerTargetPixelFile(filename_tpf_one_center)`.  That will get rid of the warning without reducing the test coverage in any significant way.

barentsen commented on lightkurve/targetpixelfile.py:
```diff
@@ -1130,6 +1131,163 @@ def cutout(self, center=None, size=5):
+                        y_val_range = np.nanmax(y_vals) - np.nanmin(y_vals)
+                        x_axis_min = np.nanmin(x_vals)
+                        x_axis_max = np.nanmax(x_vals)
+                        y_axis_min = np.nanmin(y_vals) - 0.05*y_val_range
+                        y_axis_max = np.nanmax(y_vals) + 0.05*y_val_range
```
The two lines above will indirectly trigger the following warning during the tests:
```
/home/gb/dev/lightkurve/lightkurve/targetpixelfile.py:1280: UserWarning: Attempting to set identical bottom == top == 1.0 results in singular transformations; automatically expanding.
    gax.set_ylim(y_axis_min, y_axis_max)
```

This is because the unit test uses a fake TPF (`filename_tpf_one_center`) in which the pixel values are constant over time (i.e. the pixel values in that file are always zero, except for the center pixel, which is always one).

It may not be completely unusual to have TPFs with pixels that are zero all the time, so the best workaround here may be something like this:
```suggestion
                        y_axis_min = np.nanmin(y_vals) - 0.05*max(y_val_range, 0.1)
                        y_axis_max = np.nanmax(y_vals) + 0.05*max(y_val_range, 0.1)
```

barentsen commented on lightkurve/targetpixelfile.py:
```diff
@@ -1130,6 +1131,163 @@ def cutout(self, center=None, size=5):
+        ax : `~matplotlib.axes.Axes`
+            A matplotlib axes object to plot into. If no axes is provided,
+            a new one will be generated.
+        normalized : bool
```
```suggestion
        normalize : bool
```

barentsen commented on lightkurve/targetpixelfile.py:
```diff
@@ -1130,6 +1131,163 @@ def cutout(self, center=None, size=5):
+            style = MPLSTYLE 
+        
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore")
```
Complete nit-pick: it's always good to limit the types of warnings being ignored, so we don't accidentally silence e.g. an important deprecation warning:
```suggestion
            warnings.simplefilter("ignore", category=(RuntimeWarning, LightkurveWarning))
```

barentsen commented on lightkurve/targetpixelfile.py:
```diff
@@ -1130,6 +1131,163 @@ def cutout(self, center=None, size=5):
+
+                lc = self.to_lightcurve(aperture_mask=masks[j])
+
+                if normalize: # overrides corrector function
```
Let's change this into:
```suggestion
                if normalize or corrector_func is None: # overrides corrector function
```
...which would allow us to get rid of lines 1189-1191, which reduces code duplication! :+1:

barentsen commented on lightkurve/targetpixelfile.py:
```diff
@@ -1130,6 +1131,163 @@ def cutout(self, center=None, size=5):
+                        x_axis_max = np.nanmax(x_vals)
+                        y_axis_min = 0
+                        y_axis_max = max(np.nanmax(y_vals),1e-99)
+                    elif not periodogram:
```
Nitpick: this can of course just be `else`:
```suggestion
                    else:
```

barentsen commented on lightkurve/targetpixelfile.py:
```diff
@@ -1130,6 +1131,163 @@ def cutout(self, center=None, size=5):
+                        gax.set_facecolor(tpf_plot.cmap(tpf_plot.norm(self.flux[0,x,y])))
+                        if periodogram:
+                            gax.plot(x_vals, y_vals, 'w-', lw=0.5)
+                        elif not periodogram:
```
```suggestion
                        else:
```

barentsen commented:
@astrobel Oops, I am getting blank plots by default, unless I use `show_flux=True`.  Could it be that the plot is showing white markers on a white background by default?

Example:

![pixelplot](https://user-images.githubusercontent.com/817669/87107402-87b2d500-c214-11ea-8066-f1293faf5139.png)

astrobel commented:
> @astrobel Oops, I am getting blank plots by default, unless I use `show_flux=True`. Could it be that the plot is showing white markers on a white background by default?

Yes, looks like that was just a typo on my end! Should be fixed now.

barentsen commented:
Thanks again @astrobel!

I took the liberty to rebase your PR on the `master` branch, because there have been a few changes there which affect this PR. Most notably, `tpf.flux` is now an AstroPy quantity, which was causing the tests to fail.

In addition, I spotted a few more ways to simplify this function:
* I removed the logic related to setting the x and y limits.  I found that we can achieve identical behavior by calling `gax.margins(y=.1, x=0)`.  This tells matplotlib to  auto-scale the X limits to the min-max range, and the Y limits to the min-max range plus 10%.
* I removed the `normalize` option.  Given that we don't show Y axis ticks, and given that matplotlib auto-scales the Y limits on each subplot, I found that setting `normalize` to either True or False didn't actually change the plot.
* I noticed that a second helper plot was created using `imshow` for the purpose of being able to map flux values to a correct color when `show_flux=True`:
```
tpf_plot = plt.imshow(self.flux[0].value)
plt.close()
```
I simplified this by removing this helper plot, and instead create the desired colormap and normalization function directly, which is faster:
```
cmap = plt.get_cmap()
norm = plt.Normalize(vmin=np.nanmin(self.flux[0].value), vmax=np.nanmax(self.flux[0].value))
```

Let me know if you have any concerns about the changes I made.  Otherwise, I'm happy for this to be merged!

astrobel commented:
Hi @barentsen, thanks for all these changes! Everything looks good to me, and I'm happy for it to be merged too!

barentsen commented:
Thanks again @astrobel! :+1: :+1: :+1:
